File: leetcode/p00916/p00916.py
from typing import List

class Solution:
    # Time Complexity: O(n+k), where k is the length of words2
    # Space Complexity: O(1)
    def wordSubsets(self, words1: List[str], words2: List[str]) -> List[str]:
        def counter(word: str) -> List[int]:
            frequency = [0] * 26
            for ch in word:
                frequency[ord(ch)-97] += 1
            return frequency
        
        subsetMap = [0] * 26
        for word in words2:
            letters = counter(word)
            for i in range(26):
                subsetMap[i] = max(letters[i], subsetMap[i])
        
        res = []
        for word in words1:
            supersetMap = counter(word)
            isSubset = True
            for i in range(26):
                if subsetMap[i] > supersetMap[i]:
                    isSubset = False
                    break
            if isSubset:
                res.append(word)
        
        return res

if __name__ == "__main__":
    s = Solution()

    assert s.wordSubsets(
        ["amazon", "apple", "facebook", "google", "leetcode"],
        ["e", "o"]) == ["facebook", "google", "leetcode"]
    
    assert s.wordSubsets(
        ["amazon", "apple", "facebook", "google", "leetcode"],
        ["l", "e"]) == ["apple", "google", "leetcode"]

File: leetcode/p00916/p00916.go
package p00916

// Time Complexity: O(n+k), where k is the length of words2
// Space Complexity: O(1)
func wordSubsets(words1 []string, words2 []string) []string {
	subsetMap := [26]int{}
	for _, word := range words2 {
		letters := counter(word)
		for i := 0; i < 26; i++ {
			if letters[i] > subsetMap[i] {
				subsetMap[i] = letters[i]
			}
		}
	}

	res := []string{}
	for _, word := range words1 {
		supersetMap := counter(word)
		isSubset := true
		for i := 0; i < 26; i++ {
			if subsetMap[i] > supersetMap[i] {
				isSubset = false
				break
			}
		}
		if isSubset {
			res = append(res, word)
		}
	}

	return res
}

func counter(word string) [26]int {
	frequency := [26]int{}
	size := len(word)
	for i := 0; i < size; i++ {
		frequency[word[i]-'a']++
	}
	return frequency
}
